firewood/src/shale/mod.rs

```diff
@@ -208,15 +231,16 @@ impl<'a, T: Storable + Debug> Deref for ObjRef<'a, T> {
 impl<'a, T: Storable> Drop for ObjRef<'a, T> {
     fn drop(&mut self) {
         #[allow(clippy::unwrap_used)]
-        let mut inner = self.inner.take().unwrap();
-        let ptr = inner.as_ptr();
-        let mut cache = self.cache.lock();
-        match cache.pinned.remove(&ptr) {
-            Some(true) => {
-                inner.dirty = None;
-            }
-            _ => {
-                cache.cached.put(ptr, inner);
+        if let Some(mut inner) = self.inner.take() {
```

Comment:
`into_inner` is only implemented on `ObjRef<'_, Node>`, not any `T: Storable`. `into_inner` and `drop` are the only places that set `inner` as `None`. When we call `into_inner`, we are forgoing the ability to save the node to the cache.

Reply:
Yeah this is a really tricky problem. The real issue is we need better control of how ObjRef's are dropped, especially since the `drop` method currently moves the inner value into the cache. I have some ideas on how to make this a lot better.

Can you please move this change out to a different PR? The other parts I'm fine with and this part isn't necessary for the fix you're working on.